### ISSUE TITLE
2.5.13 stable compile error

### ISSUE THREAD

**Issue by yoanrigal** [paritytech/parity-ethereum#11372]
Hello,

Previously I used already compiled linux stable version. But starting with 2.5.10 when I launch parity it crashes with "illegal instruction".

Suspecting my CPU don't support some instruction I decided to build from source the 2.5.13 stable.
Building from source fails for a reason I can't figure. Below is stderr.

- **Parity Ethereum version**: 2.5.13
- **Operating system**: Linux
- **Installation**: built from source

```
sort -u /proc/crypto | grep module
module       : kernel
```

```
rustc --print cfg -C target-cpu=native
debug_assertions
target_arch="x86_64"
target_endian="little"
target_env="gnu"
target_family="unix"
target_feature="fxsr"
target_feature="sse"
target_feature="sse2"
target_feature="sse3"
target_feature="ssse3"
target_os="linux"
target_pointer_width="64"
target_vendor="unknown"
unix
```

Stderr

```
Compiling parity-ethereum v2.7.0 (/path/edited/parity-ethereum)
error: could not compile `parity-ethereum`.


Caused by:
  process didn't exit successfully: `rustc --crate-name parity parity/main.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C opt-level=3 -C lto --cfg 'feature="accounts"' --cfg 'feature="default"' --cfg 'feature="ethcore-accounts"' --cfg 'feature="final"' -C metadata=5954e3ed1cf003a5 -C extra-filename=-5954e3ed1cf003a5 --out-dir /path/edited/parity-ethereum/target/release/deps -L dependency=/path/edited/parity-ethereum/target/release/deps --extern ansi_term=/path/edited/parity-ethereum/target/release/deps/libansi_term-e8845fde78930650.rlib --extern atty=/path/edited/parity-ethereum/target/release/deps/libatty-4ac0df23eebf9a3f.rlib --extern blooms_db=/path/edited/parity-ethereum/target/release/deps/libblooms_db-e3a08a5af8bd5531.rlib --extern clap=/path/edited/parity-ethereum/target/release/deps/libclap-230685577d1c8e2c.rlib --extern cli_signer=/path/edited/parity-ethereum/target/release/deps/libcli_signer-c05ce55644110118.rlib --extern client_traits=/path/edited/parity-ethereum/target/release/deps/libclient_traits-ee791a363b7e8e6a.rlib --extern common_types=/path/edited/parity-ethereum/target/release/deps/libcommon_types-d0ed3cdbce567e36.rlib --extern ctrlc=/path/edited/parity-ethereum/target/release/deps/libctrlc-3fb729e546a8be49.rlib --extern dir=/path/edited/parity-ethereum/target/release/deps/libdir-925de83c66a7538b.rlib --extern docopt=/path/edited/parity-ethereum/target/release/deps/libdocopt-386eef92e683dea0.rlib --extern engine=/path/edited/parity-ethereum/target/release/deps/libengine-54acf321d9045930.rlib --extern ethcore=/path/edited/parity-ethereum/target/release/deps/libethcore-ef996b6087fcb22c.rlib --extern ethcore_accounts=/path/edited/parity-ethereum/target/release/deps/libethcore_accounts-22e3ce2402ccea35.rlib --extern ethcore_blockchain=/path/edited/parity-ethereum/target/release/deps/libethcore_blockchain-09ccfafc1f03667d.rlib --extern ethcore_db=/path/edited/parity-ethereum/target/release/deps/libethcore_db-e94e22fd146797b5.rlib --extern ethcore_io=/path/edited/parity-ethereum/target/release/deps/libethcore_io-a99c08234b047c99.rlib --extern ethcore_light=/path/edited/parity-ethereum/target/release/deps/libethcore_light-2b748180fa7183c9.rlib --extern ethcore_logger=/path/edited/parity-ethereum/target/release/deps/libethcore_logger-193e571b5a90c2d6.rlib --extern ethcore_miner=/path/edited/parity-ethereum/target/release/deps/libethcore_miner-0819852694f19f03.rlib --extern ethcore_network=/path/edited/parity-ethereum/target/release/deps/libethcore_network-b5c02dd5f36bb7f7.rlib --extern ethcore_private_tx=/path/edited/parity-ethereum/target/release/deps/libethcore_private_tx-3ca44775ae375407.rlib --extern ethcore_service=/path/edited/parity-ethereum/target/release/deps/libethcore_service-ac0464d0f52ef3b0.rlib --extern ethcore_sync=/path/edited/parity-ethereum/target/release/deps/libethcore_sync-eff4e156544b0e50.rlib --extern ethereum_types=/path/edited/parity-ethereum/target/release/deps/libethereum_types-6b8df9bd75a8e599.rlib --extern ethkey=/path/edited/parity-ethereum/target/release/deps/libethkey-832460364e3c6f09.rlib --extern ethstore=/path/edited/parity-ethereum/target/release/deps/libethstore-28e2abfe483d7282.rlib --extern fdlimit=/path/edited/parity-ethereum/target/release/deps/libfdlimit-45e26080da097e7a.rlib --extern futures=/path/edited/parity-ethereum/target/release/deps/libfutures-2e23616a69d295cb.rlib --extern journaldb=/path/edited/parity-ethereum/target/release/deps/libjournaldb-9242d8061b22c032.rlib --extern jsonrpc_core=/path/edited/parity-ethereum/target/release/deps/libjsonrpc_core-33a822743e100fd5.rlib --extern keccak_hash=/path/edited/parity-ethereum/target/release/deps/libkeccak_hash-f22a1e2b2bb1eeae.rlib --extern kvdb=/path/edited/parity-ethereum/target/release/deps/libkvdb-fda2219b83fd737c.rlib --extern kvdb_rocksdb=/path/edited/parity-ethereum/target/release/deps/libkvdb_rocksdb-30de7add1c6f18ed.rlib --extern log=/path/edited/parity-ethereum/target/release/deps/liblog-8ae02a91fa87e0e3.rlib --extern migration_rocksdb=/path/edited/parity-ethereum/target/release/deps/libmigration_rocksdb-e858b0ac9c01567a.rlib --extern node_filter=/path/edited/parity-ethereum/target/release/deps/libnode_filter-edcee8cf9df8e7c2.rlib --extern num_cpus=/path/edited/parity-ethereum/target/release/deps/libnum_cpus-6acb48c4fdfb0be2.rlib --extern number_prefix=/path/edited/parity-ethereum/target/release/deps/libnumber_prefix-c50075144bd5dfc8.rlib --extern panic_hook=/path/edited/parity-ethereum/target/release/deps/libpanic_hook-c5a8f5ac80df2227.rlib --extern parity_bytes=/path/edited/parity-ethereum/target/release/deps/libparity_bytes-7039350456865eb9.rlib --extern parity_crypto=/path/edited/parity-ethereum/target/release/deps/libparity_crypto-8c78c2798c653c34.rlib --extern parity_daemonize=/path/edited/parity-ethereum/target/release/deps/libparity_daemonize-f9d5d9afbe29d7b4.rlib --extern parity_ethereum=/path/edited/parity-ethereum/target/release/deps/libparity_ethereum-68461386d5116f45.rlib --extern parity_hash_fetch=/path/edited/parity-ethereum/target/release/deps/libparity_hash_fetch-d3ae5fba217d7bf1.rlib --extern parity_ipfs_api=/path/edited/parity-ethereum/target/release/deps/libparity_ipfs_api-5b67426f031dbf10.rlib --extern parity_local_store=/path/edited/parity-ethereum/target/release/deps/libparity_local_store-05bc818870cfb495.rlib --extern parity_path=/path/edited/parity-ethereum/target/release/deps/libparity_path-78e9e6fb8df062bf.rlib --extern parity_rpc=/path/edited/parity-ethereum/target/release/deps/libparity_rpc-2e7cd2eee033f84f.rlib --extern parity_runtime=/path/edited/parity-ethereum/target/release/deps/libparity_runtime-f9522b1a466bc026.rlib --extern parity_updater=/path/edited/parity-ethereum/target/release/deps/libparity_updater-4f67ae132a3ff7e0.rlib --extern parity_util_mem=/path/edited/parity-ethereum/target/release/deps/libparity_util_mem-db92c9b2b2f33939.rlib --extern parity_version=/path/edited/parity-ethereum/target/release/deps/libparity_version-cc045e85d379cbb4.rlib --extern parking_lot=/path/edited/parity-ethereum/target/release/deps/libparking_lot-4e94693c2d96fe72.rlib --extern regex=/path/edited/parity-ethereum/target/release/deps/libregex-060032b32108fbb2.rlib --extern registrar=/path/edited/parity-ethereum/target/release/deps/libregistrar-d414773229f07d26.rlib --extern rlp=/path/edited/parity-ethereum/target/release/deps/librlp-ce38b502860047a0.rlib --extern rpassword=/path/edited/parity-ethereum/target/release/deps/librpassword-83348f378bfb4c1e.rlib --extern rustc_hex=/path/edited/parity-ethereum/target/release/deps/librustc_hex-da126fe3ee1d813a.rlib --extern semver=/path/edited/parity-ethereum/target/release/deps/libsemver-ecf0bcbf993ee25f.rlib --extern serde=/path/edited/parity-ethereum/target/release/deps/libserde-00c42a9bad9ded91.rlib --extern serde_derive=/path/edited/parity-ethereum/target/release/deps/libserde_derive-777c79496241148e.so --extern serde_json=/path/edited/parity-ethereum/target/release/deps/libserde_json-bc9c9ac85b7b5615.rlib --extern snapshot=/path/edited/parity-ethereum/target/release/deps/libsnapshot-2461c50e7f43117f.rlib --extern spec=/path/edited/parity-ethereum/target/release/deps/libspec-49ea160b737126bb.rlib --extern term_size=/path/edited/parity-ethereum/target/release/deps/libterm_size-76bbc1da28817c31.rlib --extern textwrap=/path/edited/parity-ethereum/target/release/deps/libtextwrap-fe945123699d7d23.rlib --extern toml=/path/edited/parity-ethereum/target/release/deps/libtoml-f7e8cbf8176d1c6b.rlib --extern verification=/path/edited/parity-ethereum/target/release/deps/libverification-c8e4a1162e8be48f.rlib -Ctarget-feature=+aes,+sse2,+ssse3 -L native=/path/edited/parity-ethereum/target/release/build/parity-secp256k1-8b54f4b8a335769d/out -L native=/path/edited/parity-ethereum/target/release/build/parity-snappy-sys-daedefec55510dc4/out/build -L native=/path/edited/parity-ethereum/target/release/build/jemalloc-sys-037afa5bbb4ad861/out/build/lib -L native=/path/edited/parity-ethereum/target/release/build/ring-7e30ef4c5016f6ca/out -L native=/path/edited/parity-ethereum/target/release/build/backtrace-sys-b5d71adda66bf67b/out -L native=/path/edited/parity-ethereum/target/release/build/natpmp-13bcdcbfeb8681e6/out -L native=/path/edited/parity-ethereum/target/release/build/librocksdb-sys-3676f50ae02382ed/out -L native=/path/edited/parity-ethereum/target/release/build/librocksdb-sys-3676f50ae02382ed/out` (signal: 9, SIGKILL: kill)
```

**Comment by dvdplm** [comment]
Hi, yeah, you need `aes-ni` support, so if your CPU is older than ~10 years it's not going to work I'm afraid.

**Comment by niklasad1** [comment]
> Hi, yeah, you need aes-ni support, so if your CPU is older than ~10 years it's not going to work I'm afraid.

I don't think that is the problem in this case when @yoanrigal is trying to build parity from source. `RustCrypto` should detect whether `aes-ni` feature is available and then use the software implementation instead AFAIK.

It looks to me that `librocksdb-sys` fails for some reason, maybe it run out of memory? (Has happened to me)

Can you try building https://github.com/rust-rocksdb/rust-rocksdb separately or investigate why `rustc` gets killed during compilation?

**Comment by dvdplm** [comment]
It could be `rust-rocksdb` (maybe CRC32?) but my prime suspect would be `libsecp256k1` – I think recent versions require `aes-ni` and does not provide a software alternative.

So if `rust-rocksdb` compiles, I'd try the same with https://github.com/paritytech/rust-secp256k1/

**Comment by yoanrigal** [comment]
Building separately worked, but then still the same compile error for parity-ethereum.

I also tried to build after a modification to `.cargo/config` and `scripts/gitlab/build-linux.sh` files, commenting this line:
`#rustflags = ["-Ctarget-feature=+aes,+sse2,+ssse3"]`
Still the same error.

I then checked the memory and yes I ran out of memory
`(signal: 9, SIGKILL: kill)` happened when the `rustc `process topped 50% mem usage
So I modified mem parameters on my system to allow more memory usage. In my case the process needed approx. 90% mem usage to compile successfully.
`    Finished release [optimized] target(s) in 57m 03s`

But when running the binary after compilation, I still get `Illegal instruction (core dumped)`
I give up.

**Comment by niklasad1** [comment]
That is unfortunate, sorry I missed that we had those features hardcoded in `.cargo/config`

Do you mind doing `$ cargo build --verbose &> out.log` and provide the output or grep for `aes`?

I think it is likely that other people will run into this issue too.

**Comment by yoanrigal** [comment]
Without commenting in `.cargo/config`, the output log contains `aes `references on every `rustc` run

```
...
   Compiling unicode-xid v0.2.0
     Running `rustc --crate-name build_script_build /path/edited.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.65/build.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="std"' -C metadata=edc92bacf56d652d -C extra-filename=-edc92bacf56d652d --out-dir /path/editedparity-ethereum/target/debug/build/libc-edc92bacf56d652d -L dependency=/path/editedparity-ethereum/target/debug/deps --cap-lints allow -Ctarget-feature=+aes,+sse2,+ssse3`
     Running `rustc --edition=2018 --crate-name build_script_build /path/edited.cargo/registry/src/github.com-1ecc6299db9ec823/proc-macro2-1.0.1/build.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="proc-macro"' -C metadata=5b52bab379e3f1d3 -C extra-filename=-5b52bab379e3f1d3 --out-dir /path/editedparity-ethereum/target/debug/build/proc-macro2-5b52bab379e3f1d3 -L dependency=/path/editedparity-ethereum/target/debug/deps --cap-lints allow -Ctarget-feature=+aes,+sse2,+ssse3`
     Running `rustc --crate-name unicode_xid /path/edited.cargo/registry/src/github.com-1ecc6299db9ec823/unicode-xid-0.2.0/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C debuginfo=2 --cfg 'feature="default"' -C metadata=840aeb7008349123 -C extra-filename=-840aeb7008349123 --out-dir /path/editedparity-ethereum/target/debug/deps -L dependency=/path/editedparity-ethereum/target/debug/deps --cap-lints allow -Ctarget-feature=+aes,+sse2,+ssse3`
     Running `rustc --edition=2018 --crate-name cfg_if /path/edited.cargo/registry/src/github.com-1ecc6299db9ec823/cfg-if-0.1.10/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C debuginfo=2 -C metadata=5ce6f5a306c8ebd7 -C extra-filename=-5ce6f5a306c8ebd7 --out-dir /path/editedparity-ethereum/target/debug/deps -L dependency=/path/editedparity-ethereum/target/debug/deps --cap-lints allow -Ctarget-feature=+aes,+sse2,+ssse3`
   Compiling syn v1.0.5
     Running `rustc --edition=2018 --crate-name build_script_build /path/edited.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.5/build.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C debuginfo=2 --cfg 'feature="clone-impls"' --cfg 'feature="default"' --cfg 'feature="derive"' --cfg 'feature="extra-traits"' --cfg 'feature="fold"' --cfg 'feature="full"' --cfg 'feature="parsing"' --cfg 'feature="printing"' --cfg 'feature="proc-macro"' --cfg 'feature="quote"' --cfg 'feature="visit"' -C metadata=7deafa569c8ec4a2 -C extra-filename=-7deafa569c8ec4a2 --out-dir /path/editedparity-ethereum/target/debug/build/syn-7deafa569c8ec4a2 -L dependency=/path/editedparity-ethereum/target/debug/deps --cap-lints allow -Ctarget-feature=+aes,+sse2,+ssse3`
...
```

**Comment by niklasad1** [comment]
@yoanrigal 

I meant to build parity with commenting `.cargo/config ` away. Because from my understanding that's what you before and still got `Illegal instruction`, thus I was curious to detect whether some else brought in `aes` and caused `Illegal Instruction`.

With `.cargo/config` settings then `aes` will always be enabled even if your platform doesn't support it.

I'm open to remove the `aes` hardcoded stuff in `.cargo/config` if we can build it successfully (without any `"-Ctarget-feature=+aes,+sse2,+ssse3"`)

**Comment by yoanrigal** [comment]
Ok sorry.
With commenting in `.cargo/config`, the log contains no reference to `aes`
But even in this case, running the compiled binary leads to `Illegal instruction (core dumped)` which probably comes from something else I don't explain. I probably did some shit but it comes way beyond my abilities.
Anyway even if that's not the case on the system I'm using for this right now, every modern CPU support `aes`, so it should tentatively no more be an issue.

**Comment by yoanrigal** [comment]
@niklasad1 
Indeed I did some shit.
I was sync with parity `master `branch, not `stable `branch.

Now I sync `stable` branch
I commented `#rustflags = ["-Ctarget-feature=+aes,+sse2,+ssse3"]` in `scripts/gitlab/build-linux.sh` file.
Nothing related to `aes`, nothing to comment in `.cargo/config` for this `stable `branch

Having my memory optimized compile succeeded and the `parity `binary, version `2.5.13 `runs without error.
